File: 07/07.py
from itertools import permutations

def parse_file(filename):
    with open(filename) as f:
        code = f.readline()
        ints = code.split(',')
        program = [int(i) for i in ints]
        return program

def parse_instruction(instruction):
    full_instruction = '{:05d}'.format(instruction)
    return (
        instruction % 100,
        int(full_instruction[2]),
        int(full_instruction[1]),
    )

class Machine:
    def __init__(self, program, phase):
        self.tape = program[:]
        self.position = 0
        self.inputs = [phase]
        self.output = None

    def read(self, delta=0, mode=1):
        if mode == 1:
            return self.tape[self.position + delta]
        if mode == 0:
            return self.tape[self.tape[self.position + delta]]

    def write(self, delta, value):
        pos = self.read(delta)
        self.tape[pos] = value

    def input(self):
        return self.inputs.pop(0)

    def run(self):
        while True:
            instruction = self.read()
            op_code, mode_a, mode_b = parse_instruction(instruction)

            if op_code == 99:
                return self.output, True

            if op_code == 1:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                self.write(3, a + b)
                self.position += 4

            if op_code == 2:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                self.write(3, a * b)
                self.position += 4

            if op_code == 3:
                self.write(1, self.input())
                self.position += 2

            if op_code == 4:
                self.output = self.read(1, mode_a)
                self.position += 2
                return self.output, False

            if op_code == 5:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                if a != 0:
                    self.position = b
                else:
                    self.position += 3

            if op_code == 6:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                if a == 0:
                    self.position = b
                else:
                    self.position += 3

            if op_code == 7:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                value = 1 if a < b else 0
                self.write(3, value)
                self.position += 4

            if op_code == 8:
                a = self.read(1, mode_a)
                b = self.read(2, mode_b)
                value = 1 if a == b else 0
                self.write(3, value)
                self.position += 4

def run_amps(program, phases):
    amps = [Machine(program, phase) for phase in phases]
    value = 0
    done = False
    while not done:
        for amp in amps:
            amp.inputs.append(value)
            value, done = amp.run()
    return value

def find_max(program, phases):
    max_value = 0
    max_phase = []
    for phases in permutations(phases):
        output = run_amps(program, phases)
        if output > max_value:
            max_value = output
            max_phase = phases

    return max_value

program = parse_file('input.amp')

print 'Part 1:', find_max(program, [0, 1, 2, 3, 4])
print 'Part 2:', find_max(program, [5, 6, 7, 8, 9])

File: 07/07.clj
(ns dec02
    (:require [clojure.string :as str])
    (:require [clojure.math.combinatorics :as combo]))

(defn parse-file [file-name]
    (vec (map read-string (str/split (slurp file-name) #","))))

(defn parse-instruction [n]
  [(mod n 100)
   (int (/ (mod n 1000) 100))
   (int (/ (mod n 10000) 1000))
   (int (/ (mod n 100000) 10000))])

(defn positive-position [_ position] (pos? position))

(defn read-from [program position mode]
  (case mode
    0 (nth program (nth program position))
    1 (nth program position)))

(defn write-to [program position value]
  (let [pos (nth program position)]
    (assoc program pos value)))

(defn run [program position inputs]
    (loop [program program
           position position
           output (first inputs)
           inputs inputs]
        (let [[op-code mode-a mode-b mode-c]
                (parse-instruction (read-from program position 1))
              [program' position' output' inputs' done]
            (case op-code
                99 [program position output inputs true]

                1 [(write-to program (+ position 3)
                        (+
                        (read-from program (+ position 1) mode-a)
                        (read-from program (+ position 2) mode-b)))
                    (+ position 4)
                    output inputs false]

                2 [(write-to program (+ position 3)
                        (*
                        (read-from program (+ position 1) mode-a)
                        (read-from program (+ position 2) mode-b)))
                    (+ position 4)
                    output inputs false]

                3 [(write-to program (+ position 1) (first inputs))
                    (+ position 2)
                    output (rest inputs) false]

                4 [program
                    (+ position 2)
                    (read-from program (+ position 1) mode-a)
                    inputs true]

                5 [program
                    (if (= (read-from program (+ position 1) mode-a) 0)
                        (+ position 3)
                        (read-from program (+ position 2) mode-b))
                    output inputs false]

                6 [program
                    (if (= (read-from program (+ position 1) mode-a) 0)
                        (read-from program (+ position 2) mode-b)
                        (+ position 3))
                    output inputs false]

                7 [(write-to program (+ position 3)
                        (if (<
                            (read-from program (+ position 1) mode-a)
                            (read-from program (+ position 2) mode-b))
                        1 0))
                    (+ position 4)
                    output inputs false]

                8 [(write-to program (+ position 3)
                        (if (=
                            (read-from program (+ position 1) mode-a)
                            (read-from program (+ position 2) mode-b))
                        1 0))
                    (+ position 4)
                    output inputs false])]
            (if done
                {:program program'
                 :position position'
                 :output output'
                 :done (= op-code 99)}
                (recur program' position' output' inputs')))))

(defn run-amps [program phases]
    (loop [s0 (run program 0 [(nth phases 0) 0])
           s1 (run program 0 [(nth phases 1) (:output s0)])
           s2 (run program 0 [(nth phases 2) (:output s1)])
           s3 (run program 0 [(nth phases 3) (:output s2)])
           s4 (run program 0 [(nth phases 4) (:output s3)])]
        (if (:done s4)
            (:output s4)
            (let [s0' (run (:program s0) (:position s0) [(:output s4)])
                  s1' (run (:program s1) (:position s1) [(:output s0')])
                  s2' (run (:program s2) (:position s2) [(:output s1')])
                  s3' (run (:program s3) (:position s3) [(:output s2')])
                  s4' (run (:program s4) (:position s4) [(:output s3')])]
                (recur s0' s1' s2' s3' s4')))))

(defn find-max [program phases]
    (->> (combo/permutations phases)
         (map #(run-amps program %))
         (reduce max)))

(let [program (parse-file "input.amp")]
    (println "Part 1:" (find-max program [0 1 2 3 4]))
    (println "Part 2:" (find-max program [5 6 7 8 9])))
